### ISSUE TITLE
Rule proposal: Add strict-comparisons rule

### ISSUE THREAD

**Issue by longngn** [typescript-eslint/typescript-eslint#5647]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Many developers mistakenly use `===` or `!==` to compare objects whereas they intend to use `isEqual` or `.equals`. The same for strings where most will want to use `localeCompare` instead of operators `<`, `<=`, `>=`, `>`.

I would like to port strict-comparisons rule from TSLint to here: https://palantir.github.io/tslint/rules/strict-comparisons/

### Fail Cases

```typescript
const s1 = 'a'
const s2 = 'b'
if (a < b) {}

const o1 = {}
const o2 = {}
if (o1 === o2) {}
```

### Pass Cases

```typescript
const s1 = 'a'
const s2 = 'b'
if (a.localeCompare(b) < 0) {}

const o1 = {}
const o2 = {}
if (o1 === o2) {}
```

**Comment by bradzacher** [comment]
I don't think I agree with the premise of both of these checks.

Comparing strings using `>` and `<` is fine for a lot of cases if all you want is a *consistent* ordering for a list of strings. If you're instead looking for a "locale correct" ordering of strings, then it would be better to use `localeCompare`.
But there are a number of cases where it's completely fine to use one or the other - so a rule that always enforces `localeCompare` doesn't seem entirely correct.

For objects similarly - there are many cases where comparing by reference is what you really want to do. One case in point is when you've got readonly object types - comparing by reference is exactly what you want to do.
Even with mutable objects, there are a lot of cases where a reference check is more than adequate in determining some degree of equality.

The point I'm getting at is that the code that this rule enforces is not always going to be the correct thing to do - not just at a "some people mightn't like this", but even within a single codebase you might need to both use and not use the rule.

Looking at the history - the rule was a very late addition to tslint, which is why it's not covered in our [rule alternative guide](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/TSLINT_RULE_ALTERNATIVES.md). This is the first request for the rule in 3 years, so I don't think it was a highly used rule, which doesn't indicate that it is useful for the broader TS community.

I'd be interested in hearing @JoshuaKGoldberg's thoughts given he was involved in the original TSLint implementation.

**Comment by JoshuaKGoldberg** [comment]
Yeah I'm not too motivated to add a rule that bans string `<`s or object `===`s. We don't want to be in the business of yelling at people to use more verbose code when it doesn't likely matter. In retrospect from TSLint, I don't think those features of the rule were particularly useful for almost any users. +1 to this not being a particularly useful rule.

The one case for this rule that I can see being useful is object greater than / less than equality:

https://www.typescriptlang.org/play?#code/MYewdgzgLgBCBGArApsKBGGBeGBvAvgNwBQoksCKaATNnkcQJYBmMAFJahjADxxJdqASnpA

```ts
const object1 = {};
const object2 = {};
if (object1 < object2) {}
```

Maybe we could accept a trimmed down version of the rule that just flags object greater-than / less-than checks?

**Comment by bradzacher** [comment]
One big issue I see is also the most common mistake we see when people look at TS types.

`{}` does not mean "an object" - it means "any non-null value".
Put another way:
```ts
function compare(a: {}, b: {}) {
  return a < b;
}

compare({}, {}); // obviously bad
compare(() => {}, () => {}); // obviously bad

compare(1, 2); // completely fine
compare('a', 'b'); // mostly fine
```
[playground](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABBOBbADgQwE4FMAUmAXIgN4C+ANIgEYkUCUZAUIonlCNkpogDy0A3M3LNmKDDgIVqjQYgD0CxHBoA3eCADOAGwCetTABNxaLHnz4mAXgB8ZKoiuI7DhvKUr1m3QZrGxCXMCAEZqACZ3RWUgnVwoXH1EYBgwXFNJCwByTCzqLJosqM9UOC0oJJS0oA)

So the whole premise of disallowing `(type {}) === (type {})` doesn't actually make sense in TS because it doesn't mean "an object".

Going one step further to prove my point - even a type that looks like an object still doesn't mean an object:
```ts
function compare(a: {toString(): string}, b: {toString(): string}) {
  return a < b;
}

compare({}, {}); // obviously bad
compare(() => {}, () => {}); // obviously bad

compare(1, 2); // completely fine
compare('a', 'b'); // mostly fine
```
[playground](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABBOBbADgQwE4FMAUmAXIgN5RwDKU2MYA5vgJQkDONd9AvgDSIBGJclQ4NmbUdyZkAUIkR4oIbEkyIAPAIDcMrjJkoMOAqV5kuTLYgD01xHH4A3eCFYAbAJ4DMAEwNosPHxmRABeAD5zPhCI80sbOwdnOFdPbz9-IyCARj4AJnjbZAC3XChcNOA6XEzAggByTHq+ev56wrtUOHZK6qA)

So really the only way this rule is safe and correct is if we ensure that for `a < b`, both `a` and `b` are not super-types of `number`. Which is a difficult thing to exactly prove given TS doesn't expose any type relationship APIs.

So I'm not sure there's any part of the rule that we are strongly in favour of.

**Comment by JoshuaKGoldberg** [comment]
🙃 another victim of https://github.com/microsoft/TypeScript/issues/9879. Marking as blocked until that one's resolved.

**Comment by neongreen** [comment]
FWIW I'd love to have this rule available for my own code. I _can_ think of situations where I might want reference equality, but those are rare and I'd very much rather see a warning for it that I'd have to explicitly turn off.

I'm not saying that people should be pushed towards enabling this rule — far from it. But I would appreciate having it.

**Comment by mike-hogan** [comment]
Here is why I would like something like what this feature request is driving at.  

I use tiny types all the time:

```
export interface ValueType<T> {
    value: T;
}

export interface CustomerId extends ValueType<string> {
    _type: 'customer.id';
}

export function customerId(value: string): CustomerId {
    return {
        _type: 'customer.id',
        value,
    };
}

const id1 = customerId("1")
const id2 = customerId("2")

if(id1 === id2) // I want to guard against this

if(id.value === id.value) // in favour of this
```

**Comment by zackdotcomputer** [comment]
I agree that this rule should not be applied by default, but I think there are valid, common mistakes that catch folks out and that having the ability to enable this rule would be helpful for preventing inconsistencies and errors.

Specifically, we were just caught out by the gotcha around `Date === Date` checks. So I went looking for the rule to enable to ban `Date === Date` checks to prevent members of our team from making that error in the future. It would be useful if I could put a warning on this check and others like it without having to make a custom rule.

**Comment by pbklink** [comment]
Have to convert some code which previously used `number` to `Decimal` (decimal.js).  Unfortunately the equivalence and comparisons all still happily compile but are now wrong.  So need to carefully go through all the code to fix.  This rule would have really helped.

This rule would in general be extremely helpful when working with `Decimal` like types as it is so easy to accidentally enter (say) `x > y` instead of `x.greaterThan(y)`, because you are mentally still dealing with numbers.

(I also need to fix the assignment by reference instead of value difference.  However I am sure that asking for such a rule would be just plain rude :smiley:)

**Comment by JoshuaKGoldberg** [comment]
See also discussion in #9335.

**Comment by kirkwaiblinger** [comment]
> See also discussion in #9335.

Coming from that issue, sounds like we have consensus to proceed with banning the `{ a: 0 } < { a: 0 }` case, at a minimum. Marking accepting PRs for _that subset_ of the proposal. 

Special-casing/options around `Date`s/`string`s _do not_ seem to have sufficient consensus to proceed at this time (alas!). 

It also sounds like we have consensus amongst team members to _reject_ the `object === object` check being part of the initial implementation of this rule 

Note - the `Date`/`string`/`object === object` cases may be still reasonable to evaluate for community engagement once the rule exists in its minimal version, but let's bring that minimal version into existence first.

***

If other team members disagree with this path forward, please yell! 🙂

**Comment by Josh-Cena** [comment]
My thinking is that the rule should by default check all cases, with `allowedForEquality` and `allowedForOrdering` as options.

**Comment by JoshuaKGoldberg** [comment]
Now that v8 is released, per https://github.com/typescript-eslint/typescript-eslint/discussions/7936, we can mark this issue as (now fully) unblocked! 🚀

**Comment by Lonli-Lokli** [comment]
I wish to have this rule - we have date trings and for them obvsl pure > < means nothing

**Comment by JoshuaKGoldberg** [comment]
@Lonli-Lokli Great! Please send a pull request 🙂 this has been marked as accepting PRs for about a year and a half.

**Comment by vekexasia** [comment]
I've implemented this exact rule as an independent ESLint plugin: **[eslint-plugin-no-object-comparison](https://github.com/vekexasia/eslint-plugin-no-object-comparison)** 

The rule discussed here in this topic is `no-object-comparison/object-equality`
you can read [more here](https://github.com/vekexasia/eslint-plugin-no-object-comparison/blob/master/docs/rules/object-equality.md) but in a nutshell:

Disallows:

```typescript
const obj1 = { x: 1 };
const obj2 = { x: 1 };
if (obj1 === obj2) { }  // Error: compares references

const arr1 = [1, 2, 3];
const arr2 = [1, 2, 3];
if (arr1 === arr2) { }  // Error: compares references

const d1 = new Date('2024-01-01');
const d2 = new Date('2024-01-01');
if (d1 === d2) { }      // Error: compares references

```

There are rules also to tweak [object comparison](https://github.com/vekexasia/eslint-plugin-no-object-comparison/blob/master/docs/rules/object-relational.md) (`<`, `>`, `<=`, `>=`). For example:

```typescript

const cash = new Money(100)
const bank = new BankAccount(101)

// how to handle this? 
if (cash < bank) {
  bank.withdraw(...)
}
```

Note this above should be allowed only if both Money and BankAccount either implement [toPrimitive](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toPrimitive) or [valueOf](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/valueOf)


Other rules that are not directly discussed here in this issue but might come **very**handy as I stumbled upon myself on those multiple times are:

- preventing `arr.includes({ id: 1})`
- preventing Collections reference lookups. Eg: `map.get({id: 1})` or `set.has({idx}`
- same of ^^ but for keys
- ...

Obviously the plugin requires type checking so be sure to follow add the appropriate configuration to your eslint config file.

I'd appreciate a star if you find it useful or feel free to open an issue if something is missing or misbehaving.

**Comment by bradzacher** [comment]
@vekexasia question: why as an independent plugin instead of PRing it into our plugin?

**Comment by vekexasia** [comment]
Hey @bradzacher you're right. That plugin has been in dev for a while now. I just pushed a subset of some rules I have been working on. 

There is no very particular reason besides being able to iterate fast.
